pkg/authentication/middlewares.go

```diff
@@ -38,29 +48,82 @@ func (m *Middleware) isAllowListed(r *http.Request) bool {
 	return false
 }
 
-func (m *Middleware) OAuth2Authentication(next http.Handler) http.Handler {
+func (m *Middleware) OAuth2AuthenticationChain() []func(http.Handler) http.Handler {
+	return []func(http.Handler) http.Handler{
+		m.OAuth2BearerAuthentication,
+		m.OAuth2CookieAuthentication,
+	}
+}
+
+func (m *Middleware) OAuth2BearerAuthentication(next http.Handler) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		ctx, span := m.tracer.Start(r.Context(), "authentication.Middleware.OAuth2Authentication")
+		ctx, span := m.tracer.Start(r.Context(), "authentication.Middleware.OAuth2BearerAuthentication")
 		defer span.End()
-		r = r.WithContext(ctx)
 
 		if m.isAllowListed(r) {
-			next.ServeHTTP(w, r.WithContext(ctx))
+			next.ServeHTTP(w, r)
 			return
 		}
 
-		rawAccessToken, err := getBearerToken(r.Header)
-		if err != nil {
-			m.unauthorizedResponse(w, err)
-			return
-		}
+		var (
+			principal *Principal
+			err       error
+		)
 
 		// add the Otel HTTP Client
 		r.WithContext(OtelHTTPClientContext(r.Context()))
 
-		principal, err := m.oauth2.Verifier().VerifyAccessToken(r.Context(), rawAccessToken)
+		if rawAccessToken, found := m.getBearerToken(r.Header); found {
+			principal, err = m.oauth2.Verifier().VerifyAccessToken(r.Context(), rawAccessToken)
+			if err != nil {
+				m.unauthorizedResponse(w, err)
+				return
+			}
+		}
+
+		ctx = PrincipalContext(ctx, principal)
+		next.ServeHTTP(w, r.WithContext(ctx))
+	})
+}
+
+func (m *Middleware) OAuth2CookieAuthentication(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx, span := m.tracer.Start(r.Context(), "authentication.Middleware.OAuth2CookieAuthentication")
+		defer span.End()
+
+		if m.isAllowListed(r) {
+			next.ServeHTTP(w, r)
+			return
+		}
+
+		err := fmt.Errorf("no authentication token found")
+
+		principal := PrincipalFromContext(r.Context())
+		if principal != nil {
+			// principal != nil means bearer authentication set the principal and we can move on
+			next.ServeHTTP(w, r)
+			return
+		}
+
+		if tokens, foundAny := m.getCookieTokens(r); foundAny &&
+			tokens.accessToken != "" && tokens.idToken != "" {
+			// we first validate accessToken, but rely on ID token validity and info to populate Principal attributes
+			_, errAccessToken := m.oauth2.Verifier().VerifyAccessToken(r.Context(), tokens.accessToken)
+			principal, err = m.oauth2.Verifier().VerifyIDToken(r.Context(), tokens.idToken)
```

Comment:
We do that because that is the intended use of the access token.
Also, with the support for API based token validation, in case of revocation, we benefit from that in case the access token gets revoked before the ID token expires, we would know immediately.

But I understand what you're saying, since we're not actually using the scopes from the access token we don't really "need" it, and we benefit from it only for when we validate the access token without the jwks